snap/validate_test.go

```diff
@@ -2220,3 +2224,74 @@ func (s *YamlSuite) TestValidateLinksValues(c *C) {
 		c.Check(err, ErrorMatches, l.err)
 	}
 }
+
+func (s *ValidateSuite) TestSimplePrereqTracker(c *C) {
+	// check that it implements the needed interface
+	var _ snapstate.PrereqTracker = SimplePrereqTracker{}
+
+	info := &Info{
+		Plugs: map[string]*PlugInfo{},
+	}
+	info.Plugs["foo"] = &PlugInfo{
+		Snap:      info,
+		Name:      "sound-themes",
+		Interface: "content",
+		Attrs:     map[string]interface{}{"default-provider": "common-themes", "content": "foo"},
+	}
+	info.Plugs["bar"] = &PlugInfo{
+		Snap:      info,
+		Name:      "visual-themes",
+		Interface: "content",
+		Attrs:     map[string]interface{}{"default-provider": "common-themes", "content": "bar"},
+	}
+	info.Plugs["baz"] = &PlugInfo{
+		Snap:      info,
+		Name:      "not-themes",
+		Interface: "content",
+		Attrs:     map[string]interface{}{"default-provider": "some-snap", "content": "baz"},
+	}
+	info.Plugs["qux"] = &PlugInfo{Snap: info, Interface: "not-content"}
+
+	repo := interfaces.NewRepository()
+
+	prqt := SimplePrereqTracker{}
+	providerContentAttrs := prqt.DefaultProviderContentAttrs(info, repo)
+	c.Check(providerContentAttrs, HasLen, 2)
+	sort.Strings(providerContentAttrs["common-themes"])
+	c.Check(providerContentAttrs["common-themes"], DeepEquals, []string{"bar", "foo"})
+	c.Check(providerContentAttrs["some-snap"], DeepEquals, []string{"baz"})
+
+	for _, i := range builtin.Interfaces() {
+		c.Assert(repo.AddInterface(i), IsNil)
+	}
+
+	slotSnap := &Info{SuggestedName: "slot-snap"}
+	barSlot := &SlotInfo{
+		Snap:      slotSnap,
+		Name:      "visual-themes",
+		Interface: "content",
+		Attrs:     map[string]interface{}{"content": "bar"},
+	}
+	err := repo.AddSlot(barSlot)
+	c.Assert(err, IsNil)
+	providerContentAttrs = prqt.DefaultProviderContentAttrs(info, repo)
+	c.Check(providerContentAttrs, HasLen, 2)
+	c.Check(providerContentAttrs["common-themes"], DeepEquals, []string{"foo"})
+	fooSlot := &SlotInfo{
+		Snap:      slotSnap,
+		Name:      "sound-themes",
+		Interface: "content",
+		Attrs:     map[string]interface{}{"content": "foo"},
+	}
+	err = repo.AddSlot(fooSlot)
+	c.Assert(err, IsNil)
+	providerContentAttrs = prqt.DefaultProviderContentAttrs(info, repo)
+	c.Check(providerContentAttrs, HasLen, 1)
+	c.Check(providerContentAttrs["common-themes"], IsNil)
```

Comment:
Adding a check here (and in the case below) for the existence of `some-snap` would maybe make this a bit more clear. I initially was wondering why it wasn't part of the result.